modules/gapi/src/streaming/gstreamer/gstreamerptr.hpp

```diff
@@ -0,0 +1,248 @@
+//
+// Copyright 2021 Intel Corporation.
+//
+// This software and the related documents are Intel copyrighted materials,
+// and your use of them is governed by the express license under which they
+// were provided to you ("License"). Unless the License provides otherwise,
+// you may not use, modify, copy, publish, distribute, disclose or transmit
+// this software or the related documents without Intel's prior written
+// permission.
+//
+// This software and the related documents are provided as is, with no
+// express or implied warranties, other than those that are expressly
+// stated in the License.
+//
+
+#ifndef OPENCV_GAPI_STREAMING_GSTREAMER_GSTREAMERPTR_HPP
+#define OPENCV_GAPI_STREAMING_GSTREAMER_GSTREAMERPTR_HPP
+
+#include <opencv2/gapi.hpp>
+
+#ifdef HAVE_GSTREAMER
+#include <gst/gst.h>
+#include <gst/video/video-frame.h>
+#endif // HAVE_GSTREAMER
+
+#include <utility>
+
+#ifdef HAVE_GSTREAMER
+
+namespace cv {
+namespace gapi {
+namespace wip {
+
+// NOTE: Code below originally belonged to OpenCV videoio module, GSafePtr<T> template class
+//       (videoio/src/cap_gstreamer.cpp). Here GSafePtr is renamed to GStreamerPtr and reworked
+//       in order to accept only strong references and not floating ones.
+template<typename T> static inline void GStreamerPtrUnrefObject(T** pPtr)
+{
+    if (pPtr)
+    {
+        gst_object_unref(G_OBJECT(*pPtr));
+        *pPtr = nullptr;
+    }
+}
+
+template<typename T> static inline void GStreamerPtrRelease(T** pPtr);
```

Comment:
is it possible to terminate compilation fail in case of bad type? not a linkage phase